File: migration-tool/models/recurrence.go
package models

type FrequencyType string

const (
	RepeatOnce    = "once"
	RepeatDaily   = "daily"
	RepeatWeekly  = "weekly"
	RepeatMonthly = "monthly"
	RepeatYearly  = "yearly"
	RepeatCustom  = "custom"
)

type IntervalUnit string

const (
	Hours  IntervalUnit = "hours"
	Days   IntervalUnit = "days"
	Weeks  IntervalUnit = "weeks"
	Months IntervalUnit = "months"
	Years  IntervalUnit = "years"
)

type RepeatOn string

const (
	Interval       RepeatOn = "interval"
	DaysOfTheWeek  RepeatOn = "days_of_the_week"
	DayOfTheMonths RepeatOn = "day_of_the_months"
)

type Frequency struct {
	Type   FrequencyType `json:"type" validate:"required" gorm:"type:varchar(9)"`
	On     RepeatOn      `json:"on" validate:"required_if=Type interval custom" gorm:"type:varchar(18);default:null"`
	Every  int           `json:"every" validate:"required_if=On interval" gorm:"type:int;default:null"`
	Unit   IntervalUnit  `json:"unit" validate:"required_if=On interval" gorm:"type:varchar(9);default:null"`
	Days   []int32       `json:"days" validate:"required_if=Type custom On days_of_the_week,dive,gte=0,lte=6" gorm:"serializer:json;type:json"`
	Months []int32       `json:"months" validate:"required_if=Type custom On day_of_the_months,dive,gte=0,lte=11" gorm:"serializer:json;type:json"`
}

File: android/app/src/main/java/com/dkhalife/tasks/model/Recurrence.kt
package com.dkhalife.tasks.model

data class Frequency(
    val type: String = "once",
    val on: String? = null,
    val every: Int? = null,
    val unit: String? = null,
    val days: List<Int>? = null,
    val months: List<Int>? = null
)

object FrequencyType {
    const val ONCE = "once"
    const val DAILY = "daily"
    const val WEEKLY = "weekly"
    const val MONTHLY = "monthly"
    const val YEARLY = "yearly"
    const val CUSTOM = "custom"
}

object RepeatOn {
    const val INTERVAL = "interval"
    const val DAYS_OF_THE_WEEK = "days_of_the_week"
    const val DAY_OF_THE_MONTHS = "day_of_the_months"
}

object IntervalUnit {
    const val HOURS = "hours"
    const val DAYS = "days"
    const val WEEKS = "weeks"
    const val MONTHS = "months"
    const val YEARS = "years"
}
